src/ty.md

```diff
@@ -17,110 +179,378 @@ tcx: TyCtxt<'tcx>
+things. We only want to do something when we reach a type. That means there may be a lot of
+`TypeFoldable` types whose implementations basically just forward to their fields’ `TypeFoldable`
+implementations. Such implementations of `TypeFoldable` tend to be pretty tedious to write by hand.
+For this reason, there is a `derive` macro that allows you to `#![derive(TypeFoldable)]`. It is
+defined
+[here](https://github.com/rust-lang/rust/blob/master/src/librustc_macros/src/type_foldable.rs).
+
+**`subst`** In the case of substitutions the [actual
+folder](https://github.com/rust-lang/rust/blob/04e69e4f4234beb4f12cc76dcc53e2cc4247a9be/src/librustc/ty/subst.rs#L467-L482)
+is going to be doing the indexing we’ve already mentioned. There we define a `Folder` and call
+`fold_with` on the `TypeFoldable` to process yourself.  Then
+[fold_ty](https://github.com/rust-lang/rust/blob/04e69e4f4234beb4f12cc76dcc53e2cc4247a9be/src/librustc/ty/subst.rs#L545-L573)
+the method that process each type it looks for a `ty::Param` and for those it replaces it for
+something from the list of substitutions, otherwise recursively process the type.  To replace it,
+calls
+[ty_for_param](https://github.com/rust-lang/rust/blob/04e69e4f4234beb4f12cc76dcc53e2cc4247a9be/src/librustc/ty/subst.rs#L589-L624)
+and all that does is index into the list of substitutions with the index of the `Param`.
+
+## Errors
+
+It is possible for the indices in `Param` to not match with what we expect. For example, the index
+could be out of bounds or it could be the index of a lifetime when we were expecting a type. These
+sorts of errors would be caught earlier in the compiler when translating from a `hir::Ty` to a
+`ty::Ty`. If they occur later, that is a compiler bug.
+
+There is a `TyKind::Error` that is produced when the user makes this sort of error. The idea is that
+we would propagate this type and suppress other errors that come up due to it so as not to overwhelm
+the user with cascading compiler error messages.
+
+## Question: Why not substitute “inside” the adt-def?
+
+The adt-def logically represents just the *name* of the struct (e.g., `Vec`). Its contents are
+always assumed to be expressed in terms of the “internal” or “self-view” — i.e., expressed in terms
+of the generics on the struct. If nothing else, this has an efficiency win:
+
+```rust,ignore
+struct MyStruct<T> {
+  ... 100s of field .. Vec<T>
+}
+
+MyStruct<A> ==> MyStruct<B>
+```
+
+in an example like this, we can subst from `MyStruct<A>` to `MyStruct<B>` (and so on) very cheaply,
+by just replacing the one reference to `A` with `B`. But if we eagerly substituted all the fields,
+that could be a lot more work.
+
+A bit more deeply, this corresponds to structs in Rust being *nominal* types — which means that they
+are defined by their *name* (and that their contents are then indexed from the definition of that
+name, and not carried along “within” the type itself).
```

Comment:
Would it be possible having set out the above to explain how to query if a ty:Ty implements a specific Trait? 
(Thank you so much for this chapter - it is like the rosetta stone)

Reply:
@gilescope Perhaps take a look at the trait-solving chapters (already exist). Does that answer your question?